ui/src/app/views/breadcrumb/component.ts

```diff
@@ -0,0 +1,111 @@
+  styleUrls: ['../../styles/buttons.css',
+    '../../styles/cards.css',
+    '../../styles/headers.css',
+    '../../styles/inputs.css',
+    './component.css']
+})
+export class BreadcrumbComponent implements OnInit, OnDestroy {
+  subscription: Subscription;
+  breadcrumbs: Breadcrumb[];
+  constructor(
+      private activatedRoute: ActivatedRoute,
+      private router: Router) {}
+
+  /**
+   * Generate a breadcrumb using the default label and url. Uses the route's
+   * paramMap to do any necessary variable replacement. For example, if we
+   * have a label value of ':wsid' as defined in a route's breadcrumb, we can
+   * do substitution with the 'wsid' value in the route's paramMap.
+   *
+   * @param {String} label
+   * @param {String} url
+   * @param {ActivatedRoute} route
+   * @returns {Breadcrumb}
+   */
+  private static makeBreadcrumb(
+      label: string,
+      url: string,
+      route: ActivatedRoute
+  ): Breadcrumb {
+    let newLabel = label;
+    // Perform variable substitution in label only if needed.
+    if (newLabel.indexOf(':') >= 0) {
+      try {
+        const paramMap = route.snapshot.paramMap;
+        for (const k of paramMap.keys) {
+          newLabel = newLabel.replace(':' + k, paramMap.get(k));
+        }
+      } catch (e) {
+        console.log(e);
+      }
+    }
+    return {
+      label: newLabel,
+      url: url
+    };
+  }
+
+  ngOnInit() {
+    this.breadcrumbs = this.buildBreadcrumbs(this.activatedRoute.root);
+
+    this.subscription = this.router.events.filter(event => event instanceof NavigationEnd)
+      .subscribe(event => {
+        this.breadcrumbs = this.buildBreadcrumbs(this.activatedRoute.root);
+      });
+  }
+
+  ngOnDestroy() {
+    this.subscription.unsubscribe();
+  }
+
+  /**
+   * Returns array of Breadcrumb objects that represent the breadcrumb trail.
+   * Derived from current route in conjunction with the overall route structure.
+   *
+   * @param {ActivatedRoute} route
+   * @param {string} url
+   * @param {Breadcrumb[]} breadcrumbs
+   * @returns {Array<Breadcrumb>}
+   */
+  private buildBreadcrumbs(route: ActivatedRoute,
+                           url: string = '',
+                           breadcrumbs: Breadcrumb[] = []): Array<Breadcrumb> {
+
+    const ROUTE_DATA_BREADCRUMB = 'breadcrumb';
+    const children: ActivatedRoute[] = route.children;
+    if (children.length === 0) {
+      return breadcrumbs;
+    }
+    for (const child of children) {
+      if (child.outlet !== PRIMARY_OUTLET) {
+        continue;
+      }
+      if (!child.snapshot.data.hasOwnProperty(ROUTE_DATA_BREADCRUMB)) {
+        return this.buildBreadcrumbs(child, url, breadcrumbs);
+      }
+      const routeURL: string = child.snapshot.url.map(segment => segment.path).join('/');
+      if (routeURL.length > 0) {
+        url += `/${routeURL}`;
+      }
+      const label = child.snapshot.data[ROUTE_DATA_BREADCRUMB];
+      if (!breadcrumbs.some(b => b.url === url)) {
```

Comment:
Could you also add a comment here? Not sure I follow in which cases you'd see two breadcrumbs/routes mapping to the same URL (perhaps for the empty string path routes?).